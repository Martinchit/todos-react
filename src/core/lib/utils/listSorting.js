export const listSortingByCompletion = (key) => (a, b) => a[key] - b[key];
