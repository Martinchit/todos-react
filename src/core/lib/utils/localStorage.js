export const retrieveToDos = () => {
  const result = localStorage.getItem(process.env.REACT_APP_TODO_LOCAL_STORAGE_KEY);
  return result ? JSON.parse(result) : null;
};

export const updateToDos = (toDos) => {
  const stringified = JSON.stringify(toDos);
  localStorage.setItem(process.env.REACT_APP_TODO_LOCAL_STORAGE_KEY, stringified);
};
