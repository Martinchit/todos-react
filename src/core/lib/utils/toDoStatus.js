export const TO_DO_STATUS = {
  ALL: 'all',
  COMPLETED: 'completed',
  ACTIVE: 'active',
};
