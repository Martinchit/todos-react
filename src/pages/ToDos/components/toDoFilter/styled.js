import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { IoCheckmarkDoneCircleSharp, IoAlertCircleSharp } from 'react-icons/io5';
import { Box } from '../../../../core/components/atoms/Box';

export const ToDoFilterContainer = styled(Box)`
  box-shadow: ${({ theme }) => `0 -5px 5px -7px ${theme.colors.black}`};
`;

export const ToDoStatsDoneIcon = styled(IoCheckmarkDoneCircleSharp)`
  color: ${({ theme }) => theme.colors.success};
  font-size: 1.5rem;
  margin-right: 3px;
`;

export const ToDoStatsActiveIcon = styled(IoAlertCircleSharp)`
  color: ${({ theme }) => theme.colors.warning};
  font-size: 1.5rem;
  margin-right: 3px;
`;

export const ToDoFilterOption = styled(Link)`
  margin: 0 3px;
  border-bottom: ${({ $selected, theme }) =>
    $selected ? `1px solid ${theme.colors.primary}` : 'none'};
  font-weight: ${({ $selected }) => ($selected ? 800 : 400)};
  color: ${({ theme }) => theme.colors.black};
  text-decoration: none;
  padding: 0 10px;
  cursor: pointer;
`;

export const CompletedToDoDeleteButton = styled.button`
  color: ${({ theme }) => theme.colors.danger};
  cursor: pointer;
  font-weight: 600;
`;
