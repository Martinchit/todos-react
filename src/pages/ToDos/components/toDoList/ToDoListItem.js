import React from 'react';
import Switch from 'react-switch';
import PropTypes from 'prop-types';
import {
  ToDoItemContainer,
  ToDoLabel,
  ToDoEditInput,
  ItemDeleteButton,
} from './styled';
import { useCloneToDo } from '../../../../core/lib/hooks/useCloneToDo';
import { ENTER_KEY } from '../../../../core/lib/utils/inputEventKey';

export const ToDoListItem = ({ toDo, onToDoUpdate, onToDoDelete }) => {
  const [isEditMode, setIsEditMode] = React.useState(false);
  const [toDoClone, setToDoClone] = useCloneToDo(toDo);

  const onToDoCloneUpdate = React.useCallback(
    ({ target }) => {
      setToDoClone({ ...toDoClone, title: target.value });
    },
    [toDoClone]
  );

  const toggleEditMode = React.useCallback(() => {
    setIsEditMode(!isEditMode);
  }, [isEditMode, setIsEditMode]);

  const onKeyUp = React.useCallback(
    ({ code }) => {
      if (code === ENTER_KEY) {
        onToDoUpdate(toDoClone);
        toggleEditMode();
      }
    },
    [onToDoUpdate, toDoClone]
  );

  const onInputBlur = React.useCallback(() => {
    onToDoUpdate(toDoClone);
    toggleEditMode();
  }, [onToDoUpdate, toDoClone]);

  const onToDoStatusUpdate = React.useCallback(
    (completed) => {
      onToDoUpdate({ ...toDoClone, completed });
    },
    [onToDoUpdate, toDoClone]
  );

  return (
    <ToDoItemContainer
      width="96%"
      justifyContent="space-between"
      padding="5px 2.5%"
      height="40px"
    >
      <Switch
        checked={toDo.completed}
        onChange={onToDoStatusUpdate}
        width={40}
        height={20}
      />
      {!isEditMode ? (
        <ToDoLabel onDoubleClick={toggleEditMode}>{toDoClone.title}</ToDoLabel>
      ) : (
        <ToDoEditInput
          type="text"
          value={toDoClone.title}
          onChange={onToDoCloneUpdate}
          onKeyUp={onKeyUp}
          onBlur={onInputBlur}
          autoFocus
        />
      )}
      <ItemDeleteButton onClick={() => onToDoDelete(toDo.id)} />
    </ToDoItemContainer>
  );
};

ToDoListItem.propTypes = {
  toDo: PropTypes.shape({
    title: PropTypes.string,
    id: PropTypes.string,
    completed: PropTypes.bool,
  }),
  onToDoUpdate: PropTypes.func,
  onToDoDelete: PropTypes.func,
};

ToDoListItem.defaultProps = {
  toDo: {},
  onToDoUpdate: null,
  onToDoDelete: null,
};
