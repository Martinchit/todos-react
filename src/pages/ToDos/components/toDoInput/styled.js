import styled from 'styled-components';
import { TiDeleteOutline, TiArrowSortedUp, TiArrowSortedDown } from 'react-icons/ti';

export const ToDoInputField = styled.input`
  width: 80%;
  padding: 2%;
  background-color: transparent;

  :focus,
  :active,
  :hover {
    outline: none;
  }

  ::placeholder,
  ::-webkit-input-placeholder {
    font-style: italic;
  }
  :-ms-input-placeholder {
    font-style: italic;
  }
`;

export const DeleteIcon = styled(TiDeleteOutline)`
  font-size: 1.5rem;
  margin: 0 3%;
  cursor: pointer;
  visibility: ${({ $visible }) => ($visible ? 'visible' : 'hidden')};
  color: ${({ theme }) => theme.colors.danger};
`;

export const ExpandIcon = styled(TiArrowSortedDown)`
  font-size: 1.5rem;
  margin: 0 3%;
  cursor: pointer;
  visibility: ${({ $visible }) => ($visible ? 'visible' : 'hidden')};
  color: ${({ theme }) => theme.colors.primary};
`;

export const HideIcon = styled(TiArrowSortedUp)`
  font-size: 1.5rem;
  margin: 0 3%;
  cursor: pointer;
  visibility: ${({ $visible }) => ($visible ? 'visible' : 'hidden')};
  color: ${({ theme }) => theme.colors.black};
`;
