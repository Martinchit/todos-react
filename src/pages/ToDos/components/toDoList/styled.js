import styled from 'styled-components';
import { RiDeleteBinLine } from 'react-icons/ri';
import { Box } from '../../../../core/components/atoms/Box';

export const ToDoItemContainer = styled(Box)`
  box-shadow: ${({ theme }) => `0 -5px 5px -7px ${theme.colors.black}`};
`;

export const ToDoLabel = styled.label`
  width: 80%;
  cursor: pointer;
`;

export const ToDoEditInput = styled.input`
  height: 100%;
  width: 86%;
  background-color: transparent;

  :focus,
  :active,
  :hover {
    outline: none;
  }

  :focus {
    border-bottom: ${({ theme }) => `2px solid ${theme.colors.success}`};
  }
`;

export const ItemDeleteButton = styled(RiDeleteBinLine)`
  font-size: 1.5rem;
  cursor: pointer;
  color: ${({ theme }) => theme.colors.danger};
`;
