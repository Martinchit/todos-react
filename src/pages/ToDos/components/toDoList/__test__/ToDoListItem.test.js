import React from "react";
import { shallow } from 'enzyme'; 
import Switch from 'react-switch';
import 'jest-styled-components';

import { ToDoListItem } from "../ToDoListItem";
import { ToDoLabel, ToDoEditInput, ItemDeleteButton } from '../styled';
import { ENTER_KEY } from '../../../../../core/lib/utils/inputEventKey';

describe("ToDoListItem", () => {
  const toDo = {
    id: '1',
    title: 'delectus aut autem',
    completed: true
  }
  const onToDoUpdate = jest.fn();
  const onToDoDelete = jest.fn();

  it("renders correctly", () => {
    const wrapper = shallow(<ToDoListItem toDo={toDo} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('handles onToDoCloneUpdate', () => {
    const value = 'test'
    const wrapper = shallow(<ToDoListItem toDo={toDo} />)
    wrapper.find(ToDoLabel).at(0).simulate('doubleClick')
    wrapper.find(ToDoEditInput).at(0).simulate('change', { target: { value } })
    expect(wrapper.find(ToDoEditInput).at(0).props().value).toBe(value)
  })

  it('handles onToDoUpdate', () => {
    const wrapper = shallow(<ToDoListItem toDo={toDo} onToDoUpdate={onToDoUpdate} />)
    wrapper.find(ToDoLabel).at(0).simulate('doubleClick')
    wrapper.find(ToDoEditInput).at(0).props().onKeyUp({ code: 'test' });
    expect(onToDoUpdate).not.toHaveBeenCalled();
    wrapper.find(ToDoEditInput).at(0).props().onKeyUp({ code: ENTER_KEY });
    expect(onToDoUpdate).toHaveBeenCalled();
    expect(wrapper.find(ToDoEditInput).length).toBe(0)
    onToDoUpdate.mockClear();
  })

  it('handles onInputBlur', () => {
    const wrapper = shallow(<ToDoListItem toDo={toDo} onToDoUpdate={onToDoUpdate} />)
    wrapper.find(ToDoLabel).at(0).simulate('doubleClick')
    wrapper.find(ToDoEditInput).at(0).props().onBlur();
    expect(onToDoUpdate).toHaveBeenCalled();
    expect(wrapper.find(ToDoEditInput).length).toBe(0);
    onToDoUpdate.mockClear();
  })

  it('handles onToDoStatusUpdate', () => {
    const wrapper = shallow(<ToDoListItem toDo={toDo} onToDoUpdate={onToDoUpdate} />)
    wrapper.find(Switch).at(0).simulate('change')
    expect(onToDoUpdate).toHaveBeenCalled();
    onToDoUpdate.mockClear();
  })

  it('handles onToDoDelete', () => {
    const wrapper = shallow(<ToDoListItem toDo={toDo} onToDoDelete={onToDoDelete} />)
    wrapper.find(ItemDeleteButton).at(0).simulate('click')
    expect(onToDoDelete).toHaveBeenCalled();
    onToDoDelete.mockClear();
  })
})