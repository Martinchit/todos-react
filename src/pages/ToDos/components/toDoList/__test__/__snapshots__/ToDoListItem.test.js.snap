// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoListItem renders correctly 1`] = `ShallowWrapper {}`;
